### ISSUE TITLE
Default values should be coerced to strings for parity with env vars

### ISSUE THREAD

**Issue by rhys-vdw** [eval/envied#32]
```
variable :DURATION, :integer, default: 5.seconds
```

In code with default:

```
puts ENVied.DURATION.ago
# => Fri, 10 Feb 2017 00:52:54 UTC +00:00
```

With environment variable integer:

```
puts ENVied.duration.ago
# NoMethodError: undefined method `ago' for 5:Fixnum
```

Solution:

Coerce all default values to string and have them be parsed via the production method.

**Comment by javierjulio** [comment]
I don't want to say you're wrong so I think the example isn't very good because I don't understand why you would want to do this. When you do `5.seconds` that returns an object, not a simple value like an integer. If we you run `5.seconds.class` in a console you'll get an `ActiveSupport::Duration` object. You would have to call `to_i` to get the simple value that the line of code represents or `in_milliseconds`. Also its misleading because I don't know if your intention is to actually have a date or number of seconds? Maybe the better way to do it would be:

```ruby
variable :DURATION, :integer, default: 5
```

And then now you can do:

```ruby
puts ENVied.DURATION
# => 5
puts ENVied.DURATION.seconds
# => 5 seconds
puts ENVied.DURATION.seconds.ago
# => Fri, 10 Feb 2017 01:57:14 UTC +00:00
```

**Comment by rhys-vdw** [comment]
> the example is bad

Yep. The real code I used was `10.minutes`, which was just a readability thing. Of course this could apply to any value type.

```
10.seconds == 10
=> true
```

>  I think the better way to do it [...]

Yes, I did something dumb. Because what I did is wrong it should have raised an error because I treated it as the Ruby value (rather than the coerced string).

The point is that in production all environment variables are strings, so whatever is provided to the `default` should be converted to a string before it gets run through the development code (which should be as close to production as possible).

> We would have to call to_i to get the simple value that the line of code represents [...]

I think they should be first converted to strings so that they mirror the production behaviour. All environment variables are strings, so they're no doubt already being coersed with `to_i` or equivalent.

```
[5] pry(main)> 10.seconds.to_s.to_i
=> 10
```

> or call something like in_milliseconds

FYI:

```
[4] pry(main)> 10.seconds == 10
=> true
```

But this applies to all types. eg:

```
variable SOME_OBJECT, :string, default: { a: 5 }
# No error raised in development if dev forgets to call `JSON.parse`.
```

**Comment by javierjulio** [comment]
Yes I figured this is what you were getting at because recently I had to set another default and I had to do a double take as for a second I had thought, wait do I set this as a string or as the actual data type. In the case of it being a number, actually using an integer. Reviewing the README [in the Defaults section](https://github.com/eval/envied#defaults) does show the boolean default being set as a string. I've always done it as its native value. The next example though is an integer but using a proc which does use an integer rather than string representation. Interesting. There is definitely two approaches here because the [two examples linked in the More Examples section](https://github.com/eval/envied#more-examples), the first uses strings for all data types but the second does not. It might be unintentional or possibly even to work in both ways.

**Comment by javierjulio** [comment]
It looks it was considered that defaults be set as strings but then reverted. You can see commented out code that made sure [defaults were set as string and to raise an error](https://github.com/eval/envied/blob/master/lib/envied/variable.rb#L10).

**Comment by rhys-vdw** [comment]
I think it's okay for the values to be provided as whatever literal, but I'd update that to be:

```rb
    @default = options[:default].to_s
```

That way the value will still be coerced as per normal, highlighting any inconsistencies between development and production config.

**Comment by javierjulio** [comment]
But the thing is this wouldn't be applicable for some objects, for example, hashes or arrays (I'm sure there are other cases too) can be given in their native Ruby format but when you specify them as a string per the examples in the README its a very different format. The `to_s` format doesn't map to that. I see what you are getting at and I feel it should be addressed but I would be more for consistency so I would rather have the defaults be only defined as strings so there is no confusion or encountering unexpected outcomes.

**Comment by eval** [comment]
@javierjulio the restriction that defaults should always be strings was removed when default-procs were introduced.
@rhys-vdw It's recommended to always denote defaults as strings - this is in the end what you should provide via the ENV.

Looking at your usecase it might be nice to have the ability to define your own types in the Envfile:
```ruby
register_type(:duration) do |raw_string|
  _all, count, unit = */(\d+) (\w+)/.match(s)
  Integer(count).public_send(unit)
end

# as we're always processing strings less boilerplate variant might be:
register_type(:duration, /(\d+) (\w+)/){|raw, count, unit| Integer(count).public_send(unit) }

# usage:
variable :session_timeout, :duration, default: '5 minutes'
```

This way the variable is closer to the domain (ie it's always used as AS::Duration), while coming up with a way to encode this in a string.

**Comment by javierjulio** [comment]
Thanks @eval! 🙏🏻 I was going to ask you about this today. 😄 As you can see we were confused as to the direction you were going in. What I'm thinking of doing is updating all examples to use strings for default values. Curious, why did you remove the restriction for default-procs? Couldn't we have required then that the proc just returned a string too? I'm going to make a new issue listing what I plan to do for a 1.0. I'd love your input @eval and @rhys-vdw as I want to do more but thinking that should come after that 1.0 release.

**Comment by rhys-vdw** [comment]
> Looking at your usecase it might be nice to have the ability to define your own types in the Envfile

Quite possibly. Actually I think that my example is stupid anyway. Turns out the integer monkeypatch that adds `.minutes`, `.seconds` etc is not available while ENVied is initializing anyway (before Rails has started).

The approach we ended up taking was this:

```
variable :session_timeout_in_seconds, :integer, default: (5 * 60).to_s
```

Not as fancy, but it's unambiguous. I can see the appeal for custom types though. 👍
